### DISCUSSION TITLE
Support for SINGLE or MARK joins

### DISCUSSION THREAD

**Discussion by zenazn** [duckdb/duckdb#6045] [Ideas]
I noticed that the code seems to have support for `JoinType::SINGLE` and `JoinType::MARK`. I'm curious if these can be exposed in the SQL interface in some way (e.g., `t1 SINGLE JOIN t2 ON ...` or some other syntax):

https://github.com/duckdb/duckdb/blob/45fed7847f427decf24539f487c70838009444e6/src/include/duckdb/common/enums/join_type.hpp#L26-L29

I have a use case where I have several "lookup tables" that have a column or columns (the "key") that are purportedly unique. I have queries that start with an "input table" and join these "lookup tables" on their key. However, I want to preserve the invariant that every row in the "input table" has exactly one output row—that is, a "lookup table" that has two join partners should produce an error, not two output rows.

My current workaround involves performing two queries: one to produce the results, and a second to check the results for errors (I emit enough information about the join partners in the first query to do the obvious `COUNT/GROUP BY`). But `SINGLE` joins (and possibly `MARK` joins) sound like a perfect fit for this use case!

**Comment by Mytherin** [comment]
Single and mark joins are emitted as part of correlated subqueries - perhaps you could use a correlated subquery to get the result you are looking for? Specifically `SINGLE` joins are emitted as part of scalar correlated subqueries, and `MARK` joins as part of an `IN` clause (or `ANY`/`ALL`).

**Comment by zenazn** [comment]
That's helpful, but has two problems for my use case:
1. More importantly, DuckDB does not error when multiple rows match in a correlated subquery (I've filed https://github.com/duckdb/duckdb/issues/6060 because I believe this is a bug!)
2. My use case involves retrieving *multiple* output values from the "lookup table," not just the one. I could write this as multiple correlated subqueries, but the DuckDB query optimizer doesn't know how to coalesce those SINGLE joins, and doing multiple joins is less efficient (though possibly bearable). We can maybe work around this by returning a struct type, but I haven't investigated further.

I think only the first one is a blocker for my use case—if we're able to address that bug this would be a great solution for us!
